File: 017. Letter Combinations of a Phone Number/Python/Solution.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        if not digits:
            return []
        result = []
        result.append('')
        pattern = ['0', '1', 'abc', 'def', 'ghi', 'jkl', 'mno', 'pqrs', 'tuv', 'wxyz']
        for digit in digits:
            temp_result = []
            for item in result:
                for letter in pattern[ord(digit) - ord('0')]:
                    temp_result.append(item + letter)
            result = temp_result
        return result

File: 017. Letter Combinations of a Phone Number/C++/Solution.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) {
            return vector<string>();
        }
        vector<string> result;
        result.push_back("");
        vector<string> pattern{"0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        for (const auto &digit : digits) {
            vector<string> tempResult;
            for (const auto &item : result) {
                for (const auto &letter : pattern[digit - '0']) {
                    tempResult.push_back(item + letter);
                }
            }
            result = tempResult;
        }
        return result;
    }
};